### PR TITLE
Lecture 2 Solutions

### PR DESCRIPTION
**Pull request by jakalx**
### Solutions for Lecture 2

Hi @chshersh, I've completed lecture 2 :) It's been very nice and informative,
some of my earlier solutions had issues on infinite input or some corner cases,
so thanks for the great tests!

I've also added some tests regarding the 'dragonFight', I hope that's ok.

Looking forward to getting your feedback!

Sincerely,
Alex

### CODE REVIEW

File: src/Lecture2.hs [hunk 1]
```diff
@@ -52,7 +68,9 @@ zero, you can stop calculating product and return 0 immediately.
 84
 -}
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
+lazyProduct = foldr step (1 :: Int)
+  where step :: Int -> Int -> Int
+        step x y = if x == 0 then 0 else x*y
```
Comment (chshersh) [inline src/Lecture2.hs:73]: This is a neat implementation! I see that it indeed does stop correctly on zero 🙂 

It contains a space leak still but I cover this in Lecture 4 only

File: src/Lecture2.hs [hunk 2]
```diff
@@ -62,7 +80,7 @@ lazyProduct = error "TODO"
 "ccaabb"
 -}
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate = concat . fmap (take 2 . repeat)
```
Comment (chshersh) [inline src/Lecture2.hs]: That's a correct solution 👍🏻 You can even use the standard function `concatMap` to replace `concat . fmap` and `replicate 2` to replace `take 2. repeat`.

However, even in that case, the implementation allocates temporary lists inside and appends them with `++`.

It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible.

File: src/Lecture2.hs [hunk 3]
```diff
@@ -74,7 +92,12 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt n xs
+  | n < 0 = (Nothing, xs)
+  | otherwise = case splitAt n xs of
+      (h, []) -> (Nothing, h)
+      (h, (x:rest)) -> (Just x, h ++ rest)
```
Comment (chshersh) [inline src/Lecture2.hs:102]: Perfect 💎

File: src/Lecture2.hs [hunk 4]
```diff
@@ -185,7 +273,10 @@ False
 True
 -}
 isIncreasing :: [Int] -> Bool
-isIncreasing = error "TODO"
+isIncreasing [] = True
+isIncreasing (_:[]) = True
```
Comment (chshersh) [inline src/Lecture2.hs]: You can simplify the pattern here a bit 🙂 

```suggestion
isIncreasing [_] = True
```

File: src/Lecture2.hs [hunk 5]
```diff
@@ -198,7 +289,11 @@ verify that.
 [1,2,3,4,7]
 -}
 merge :: [Int] -> [Int] -> [Int]
-merge = error "TODO"
+merge xs [] = xs
+merge [] ys = ys
+merge (x:xs) (y:ys)
+  | x < y = x : merge xs (y:ys)
+  | otherwise = y : merge (x:xs) ys
```
Comment (chshersh) [inline src/Lecture2.hs]: Absolutely correct solution! 💎 
You can make it more performant by checking if both heads of the given lists are equal. In that case, you're able to reduce the size of each list by 1 on each recursive call instead of reducing the size of only a single list ✂️

File: src/Lecture2.hs [hunk 6]
```diff
@@ -215,8 +310,11 @@ The algorithm of merge sort is the following:
 [1,2,3]
 -}
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
-
+mergeSort [] = []
+mergeSort [x] = [x]
+mergeSort xs = merge (mergeSort left) (mergeSort right)
+  where (left, right) = splitAt pivot xs
+        pivot = length xs `div` 2
```
Comment (chshersh) [inline src/Lecture2.hs]: You can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Alternatively, you can implement a bottom-up approach by converting a list of elements to a list of of singleton lists and merging them pairwise until only one list is left.

File: src/Lecture2.hs [hunk 7]
```diff
@@ -288,8 +393,30 @@ It also can be:
...
+literal.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding e = case pullOutConstant e of
+  (0, Just x) -> x
+  (n, Nothing) -> Lit n
+  (n, Just x) -> Add (Lit n) x
+
+pullOutConstant :: Expr -> (Int, Maybe Expr)
```
Comment (chshersh) [inline src/Lecture2.hs:413]: You can simplify the implementation if this function if you return `(Int, [String])` instead. And you'll only need to combine variable names to the nested addition in the end

**Review by chshersh — APPROVED** [review]
Huge congratulations! You've completed the second homework 👏🏼 
This homework has some rather difficult exercises but you managed to overcome these challenges! 🥳
Keep going! The first and the most difficult half of the course is done 💪

Also, really nice implementation of the Dragon Fight task 😎
